### ISSUE TITLE
CockroachDB support

### ISSUE THREAD

**Issue by joscdk** [vapor/postgres-kit#85]
We should see if our PostgreSQL driver, and Fluent supports CockroachDB.

CockroachDB is a fairly new psql compatible database that is built to run in Clusters.

More info: https://www.cockroachlabs.com

**Comment by wafflespeanut** [comment]
Hi @joscdk! I tried using this library for CockroachDB for a project about 2 months back (RC 2.2.1), and I was quite surprised about the results. It worked fine for most of the queries out of the box. So, I decided to get back into it once my project's APIs became stable enough. Today, I noticed the version bump, and bumped it to RC 4.2.1, and now, I'm getting an error for all queries.

Here's what I did:

- I [generated the certs and ran a single-node cluster](https://www.cockroachlabs.com/docs/v2.0/secure-a-cluster.html).
- Created `test_database`
- And, ran this program:

``` swift
import NIO
import NIOOpenSSL
import PostgreSQL

let authorityCertificatePath = "/home/wafflespeanut/certs/ca.crt"
let clientPrivateKeyPath = "/home/wafflespeanut/certs/client.root.pem"
let clientCertificatePath = "/home/wafflespeanut/certs/client.root.crt"

var eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
let privateKey = OpenSSLPrivateKeySource.file(clientPrivateKeyPath)
let rootCert = OpenSSLTrustRoots.file(authorityCertificatePath)
let clientCert = OpenSSLCertificateSource.file(clientCertificatePath)
let tlsConfig = TLSConfiguration.forClient(certificateVerification: .fullVerification,
                                           trustRoots: rootCert,
                                           certificateChain: [clientCert],
                                           privateKey: privateKey)
let transportConfig = PostgreSQLConnection.TransportConfig.customTLS(tlsConfig)
let config = try! PostgreSQLDatabaseConfig(url: "postgresql://root@localhost:26257/test_database", transport: transportConfig)!
let database = PostgreSQLDatabase(config: config)

let future = database.newConnection(on: eventLoopGroup).then() { connection in
    return connection.query("CREATE TABLE IF NOT EXISTS foobar (id uuid PRIMARY KEY NOT NULL, name text NOT NULL, description text)", []) { row in
        print(row)
    }
}

try! future.wait()
```

For this, I get the following error:

```
Fatal error: 'try!' expression unexpectedly raised an error: ⚠️ PostgreSQL Error: Could not decode UInt32: 0xb3818d95 (OID).
- id: PostgreSQLError.decode
: file /home/buildnode/jenkins/workspace/oss-swift-4.1-package-linux-ubuntu-16_04/swift/stdlib/public/core/ErrorType.swift, line 184
```

I'm using **Swift 4.1.2**.

I guess the error is coming from [here](https://github.com/vapor/postgresql/blob/c34533ebf7ae34cff1430a2803c83a02bb98557a/Sources/PostgreSQL/Codable/PostgreSQLRowDecoder.swift#L80). I couldn't get any deeper, because there's been a huge rewrite over the last few versions apparently. But, I'm quite sure this worked on RC 2.2.1.

I'm unable to compile this with RC 2.2.1 now, because the package resolution confuses the the NIO versions, and I'm getting really strange build errors. Is there any way I can help with fixing this?

**Comment by wafflespeanut** [comment]
Apparently, this has been fixed in 1.0.1. I think #95 had the fix.

**Comment by joscdk** [comment]
@wafflespeanut oh that is amazing :D .. I will definitely try it out asap :D

**Comment by tanner0101** [comment]
We should consider adding a CockroachDB test to this package's [`circle.yml`](https://github.com/vapor/postgresql/blob/master/circle.yml). That way we can make sure we don't break anything going forward. :)

**Comment by ichord** [comment]
vapor postgresql 1.1.2 raise this error:

```
[psql] [2018-11-17 13:39:51 +0000] SELECT current_setting('server_version') AS "version" []
Fatal error: Error raised at top level: ⚠️ PostgreSQL Error: unknown function: current_setting()
- id: PostgreSQLError.server.error.ResolveFunction
```

for CockroachDB doesn't support `current_setting`  yet:
https://github.com/cockroachdb/cockroach/issues/26386

**Comment by ichord** [comment]
have to set `certificateVerification: .noHostnameVerification` to not verify the hostname, for postgres doesn't provide the `serverName` while initialising `OpenSSLClientHandler`

**Comment by wafflespeanut** [comment]
I've fixed that in #107 :)

**Comment by tanner0101** [comment]
I've added a `cockroach` image to this repo's `docker-compose` for testing. The package seems to work fine with CockroachDB but the tests need some slight refactoring to work with both PostgresSQL and CockroachDB. 

Cockroach runs under the `--insecure` flag in Docker and supports logging in with `root` and no password. Postgres on the other hand lets you supply a username and password to the docker image. We need a uniform way to authenticate tests connections to both DBs. 

Additionally, database selection doesn't seem to be working as expected. CockroachDB yields the following error for some test queries:

```
cockroach server: no database specified (ResolveSchema)
```

The database _should_ be getting set during the authentication request. Some digging needs to be done here to see why that isn't working and what the proper method for selecting a database in CockroachDB is.
